packages/headless/src/features/search-parameters/search-parameter-serializer.ts

```diff
@@ -12,12 +12,44 @@ function serialize(obj: SearchParameters) {
   return fragment;
 }
 
-function deserialize(fragment: string): Record<string, string> {
+function deserialize(fragment: string): SearchParameters {
   const parts = fragment.split('&').filter((part) => part.length);
-  const keyValuePairs = parts.map((part) => part.split('='));
+  const keyValuePairs = parts
+    .map((part) => part.split('='))
+    .filter(isValidPair)
+    .map(cast);
 
   return keyValuePairs.reduce((acc, pair) => {
     const [key, val] = pair;
     return {...acc, [key]: val};
   }, {});
 }
+
+function isValidPair<K extends keyof SearchParameters>(
+  pair: string[]
+): pair is [K, string] {
+  const validKey = isValidKey(pair[0]);
+  const lengthOfTwo = pair.length === 2;
+  return validKey && lengthOfTwo;
+}
+
+function isValidKey(key: string): key is keyof SearchParameters {
+  const validKeys: (keyof SearchParameters)[] = ['q', 'enableQuerySyntax'];
+  return validKeys.some((validKey) => validKey === key);
+}
+
+function cast<K extends keyof SearchParameters>(
+  pair: [K, string]
+): [K, unknown] {
+  const [key, value] = pair;
+
+  if (key === 'q') {
+    return [key, value];
+  }
+
+  if (key === 'enableQuerySyntax') {
+    return [key, value === 'true'];
```

Comment:
Here we return `true` if the string value exactly matches, and false otherwise. What do you think of this?

An alternative would be to make the logic more strict, e.g.
```
if (value === 'true') {
    return true;
}
if (value === 'false') {
    return false;
}

// else discard key and value pair.
```

Reply:
I'll have to take some more time to review this whole feature, because it's not clear to me how this fits in the bigger picture that "end users wanted to control the format of how things gets serialized, and complained that it was not the format they wanted with JSUI"

I think we're slowly falling into the same problem we had with JSUI ? 

As in, the format is "forced" on you, and too bad if you need to have another format for XYZ reason.

Reply:
Said differently:

```
 const parts = fragment.split('&').filter((part) => part.length);
  const keyValuePairs = parts.map((part) => part.split('='));	  const keyValuePairs = parts
    .map((part) => part.split('='))
    .filter(isValidPair)
    .map(cast);
```

This strongly implies that this is linked to an URL, since it expects a string, with each parameter separated by `&` and key and values linked with `=`.

Or, they would like to have the parameter `q` named `query=potato` instead. 

I guess it's simpler for us if we force that, and we'll see if we really need to have this format controllable externally, though.

Reply:
As for your original question, for the strictness of boolean : 

If "default value = false, then the strictness check needs to be for `true`".
If "default value = true, then the strictness check needs to be for `false`".

Reply:
We are offering a lot of flexibility with the current approach, and have room to add even more in a non-breaking way:

**General format**

> The format is "forced" on you

The developer does not need to use our serializer. The state of `buildSearchParameterManager` returns a key-value object. If they want to serialize it as json, use a custom serializer, store some keys and not others, they are free to do so.

**Configuring delimiters**

> This strongly implies that this is linked to an URL, since it expects a string, with each parameter separated by & and key and values linked with =

Yes, and I think url-form is a good default for the majority users of our serializer. That said, we also have a free option to make the delimiters configurable in a non-breaking way.

```
const {serializer, deserializer} = buildSearchParametersSerializer(...options...)
```
This is very similar to the [nodejs api](https://nodejs.org/api/querystring.html#querystring_querystring_stringify_obj_sep_eq_options).

I haven't added this config yet because we are still building out the base feature, and exposing such options increases the risk of someone passing an incompatible delimiter.

**Key aliases**

> They would like to have the parameter q named query=potato instead

This can be supported through another option passed when initializing the serializer. I think it is too early to add it, and should focus on building a working feature before adding configurability.

```
const keyMap: Record<keyof SearchParameters, string> = {
   q: 'query',
    sortCriteria: 'sort'
}

const {serializer, deserializer} = buildSearchParameterSerializer({keyMap});
```

**Summary (tl;dr)**

We can make the serializer configurable in a non-breaking way once the base feature is built.